src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java

```diff
@@ -249,4 +250,42 @@ public ByteBufAllocator getAllocator()
     {
         return channelConfig.getAllocator();
     }
+
+    /**
+     * Consumes bytes in the stream until the given length
+     *
+     * @param writer
+     * @param len
+     * @return
+     * @throws IOException
+     */
+    public long consumeUntil(BufferedDataOutputStreamPlus writer, long len) throws IOException
+    {
+        long copied = 0; // number of bytes copied
+        while (copied < len)
+        {
+            if (buffer.remaining() == 0)
+            {
+                try
+                {
+                    reBuffer();
+                }
+                catch (EOFException e)
+                {
+                    throw new EOFException("EOF after " + copied + " bytes out of " + len);
+                }
+                if (buffer.remaining() == 0)
+                    return copied == 0 ? -1 : copied;
```

Comment:
Heh, sorry I'm back for more. So reBuffering always does this check. I would change this to throw AssertionError because if it happens we don't actually know if it's end of stream or not because reBuffer didn't do what reBuffer is supposed to do.

Reply:
Right now the check is being made in `BigTableBlockWriter` and it throws an `IOException`, do we need to convert it into an `AssertionError`? We could convert this into an `EOFException` which is more uniform with the way the rest of the class handles things. WDYT?

Reply:
Here - https://github.com/dineshjoshi/cassandra/blob/f9573ccef63535b787e91e46fa886f3e3ff2eee8/src/java/org/apache/cassandra/io/sstable/format/big/BigTableBlockWriter.java#L213

`reBuffer()` could likely fail due to network or channel being closed by the other side. It could be treated as a normal `IOException` as we're not guaranteed that we would get those many bytes from the other end. It feels a bit weird that rest of `RebufferingByteBufDataInputPlus` throws `EOFException` and this method would throw an `AssertionError` when there may be a network issue.

Reply:
Re-reading the code, the check on [line 227](https://github.com/dineshjoshi/cassandra/blob/f9573ccef63535b787e91e46fa886f3e3ff2eee8/src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java#L277) is just being over cautious in case the `reBuffer` implementation changes and can return a empty buffer.

Reply:
If reBuffer() fails it's going to throw an IOException or RuntimeException that won't be caught there so it won't throw AssertionError. It will only throw AssertionError when reBuffer() fails to reBuffer() and also doesn't exit via an Exception. 

If reBuffer() doesn't reBuffer() and doesn't throw EOF then you don't know what's happening. That check doesn't do anything because if it doesn't read the require number of bytes it's going to throw EOF (in reBuffer()) and skip it anyways.

Reply:
Ok, makes sense. I have updated it to throw an `AssertionError`.